### DISCUSSION TITLE
Is there equivalent to webpack@5's experiments.topLevelAwait option

### DISCUSSION THREAD

**Discussion by pbadenski** [babel/babel#14657] [Q&A]
We're in the process of incrementally migrating our codebase to ESM. This can only be an incremental approach if we're allowed to import ESM modules from CJS scripts. This works when using webpack, but I can't figure out how to make it work in babel alone.

Example:
```
// test.ts - CJS Typescript
const { createFactory } = await import("./factory.mjs");
```

**Comment by JLHwung** [comment]
Are you targeting to a browser or Node.js? If you are targeting to Node.js, does https://github.com/airbnb/babel-plugin-dynamic-import-node work?

**Reply by pbadenski** [reply]
Nope, same problem - getting `SyntaxError: await is only valid in async functions and the top level bodies of modules`

**Reply by JLHwung** [reply]
Top level await has been materialized since `@babel/parser` 7.15.0. Please upgrade to latest `@babel/parser` versions. And if you meant to consume `test.ts` as a module, you should also specify `sourceType: "module"` https://babeljs.io/docs/en/options#sourcetype

If none of them helps, please share a reproduction repo so we can take a look.

**Reply by pbadenski** [reply]
Following your suggestion I created a repo to get on the same page, see: https://github.com/pbadenski/babel-14657-discussion-top-level-await

The example shows webpack targetting node, but this solution also works on a browser.

**Reply by nicolo-ribaudo** [reply]
You need a Babel config file with `sourceType: "module"` to tell Babel that those files are ESM

**Reply by pbadenski** [reply]
By "those files" I assume you mean `main.js` and `lib2.js`? That's the crux of the matter - they're not supposed to be ESM. I'm looking for a way to import ESM modules from CJS scripts.

**Reply by JLHwung** [reply]
> they're not supposed to be ESM. I'm looking for a way to import ESM modules from CJS scripts.

Top level `await import()` is invalid in commonjs. The top level await proposal enables the "module" top level as an async context, this is not applicable to the script-type sources.

There are three solutions:

1) Continue using top level `await import()`, mark the `main.js` as ESM and have Babel transforming it into a CJS file
2) Recognize `main.js` as CJS, use `require()` instead of `await import()`.
3) wrap the `await import()` into an async IIFE.

**Reply by nicolo-ribaudo** [reply]
Or you can also do something like this (`index.js`):
```js
import("./lib1.mjs").then(({ func1 }) => {
  const { func2 } = require("./lib2.js");

  func1();
  func2();
});
```

**Reply by pbadenski** [reply]
@nicolo-ribaudo thanks - yes that works, although solution is limited to scripts without further exports (I missed that scenario in my example)

@JLHwung thanks for clarification, I think I understand now. webpack@5's `experiments.topLevelAwait` handles the scenario of top level `await import()` in commonjs for the user (from what I understand by modifying the code to use the combination of transformations suggested by you and @nicolo-ribaudo). What I gather from this conversation is that there's no equivalent automatic mechanism in babel at the moment.
